src/main/host/network/queuing.rs

```diff
@@ -0,0 +1,321 @@
+    queue: QueuingDiscipline<T>,
+}
+
+impl<T: Clone + Debug + Eq + Hash> NetworkQueue<T> {
+    /// Create a new queue that will organize items according to the given `NetworkQueueKind`.
+    pub fn new(kind: NetworkQueueKind) -> Self {
+        let queue = match kind {
+            NetworkQueueKind::MinPriority => QueuingDiscipline::MinPriority(BinaryHeap::new()),
+            NetworkQueueKind::FirstInFirstOut => {
+                QueuingDiscipline::FirstInFirstOut(VecDeque::new())
+            }
+        };
+        Self {
+            membership: HashSet::new(),
+            queue,
+        }
+    }
+
+    /// Remove and drop all items from the queue.
+    pub fn clear(&mut self) {
+        match &mut self.queue {
+            QueuingDiscipline::MinPriority(heap) => heap.clear(),
+            QueuingDiscipline::FirstInFirstOut(deque) => deque.clear(),
+        }
+        self.membership.clear();
+    }
+
+    /// Returns true if the queue contains the item, and false otherwise.
+    pub fn contains(&self, item: &T) -> bool {
+        self.membership.contains(item)
+    }
+
+    /// Returns true if the queue does not contain any items.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the number of queued items.
+    pub fn len(&self) -> usize {
+        self.membership.len()
+    }
+
+    /// Returns the next available item if one exists. This function also drops the duplicate cloned
+    /// handle that was created in `NetworkQueue::try_push()`.
+    pub fn pop(&mut self) -> Option<T> {
+        match &mut self.queue {
+            QueuingDiscipline::MinPriority(heap) => heap.pop().map(|x| x.item),
+            QueuingDiscipline::FirstInFirstOut(deque) => deque.pop_front(),
+        }
+        .inspect(|x| {
+            assert!(self.membership.remove(x));
+        })
+    }
+
+    /// Pushes an item into the queue, internally cloning it to
+    /// support membership checks. It may be useful to wrap items with a `std::rc::Rc` or
+    /// `std::sync::Arc` before calling this function.
+    ///
+    /// # Panics
+    /// This function panics if the queue was configured with `NetworkQueueKind::MinPriority` and
+    /// `maybe_priority` is `None`.
+    pub fn push(&mut self, item: T, maybe_priority: Option<u64>) {
+        self.try_push(item, maybe_priority).unwrap()
+    }
+
+    /// Tries to push an item into the queue. If successful, the item is internally cloned to
+    /// support membership checks, so it may be useful to wrap items with a `std::rc::Rc` or
+    /// `std::sync::Arc` before calling this function. If unsuccessful, a `PushError` is returned
+    /// encoding the reason for the failure.
+    pub fn try_push(&mut self, item: T, maybe_priority: Option<u64>) -> Result<(), PushError> {
+        if self.contains(&item) {
+            Err(PushError::AlreadyQueued)
+        } else {
+            assert!(self.membership.insert(item.clone()));
```

Comment:
Note: this `clone()` is normally a little weird. I'd rather use an `Rc` to avoid cloning items added to the queue, but in our case we are pushing `InetSocket`s which are almost entirely covered by an `Arc` already so in our current use the clone is already cheap.

Alternatively, we could wrap items in an `Rc` internally here in the queue, but then we'd need to wrap an `Arc` around the `InetSocket` since it is an enum and not totally `Send` on its own.

So, I was thinking that for now we can just avoid the extra `Arc`s and live with the current limitation that the queue clones on `push()`, knowing the the clone is just cloning an `InetSocket` which is already designed to be cheaply cloned.

But if there is a better option here I'd love to hear it :)